(fix) O3-4363 - Add pagination to visit context switcher

chibongho opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary
<!-- Please describe what problems your PR addresses. -->
Currently, the visit context switcher only shows a set number of visits, and there is no way to select older visits pass that maximum number. This PR adds pagination support to the visit list.

## Screenshots
<!-- Required if you are making UI changes. -->
![image](https://github.com/user-attachments/assets/7db72a7e-718f-4644-8dba-73fa7849cd95)



## Related Issue
<!-- Paste the link to the Jira ticket here if one exists. -->
<!-- https://issues.openmrs.org/browse/O3- -->

## Other
<!-- Anything not covered above -->

ibacher commented:
This is a widget where I think infinite scroll makes more sense than pagination, no? Even better would probably be to be able to search by date, though.

chibongho commented:
Yeah, I wish we have better UX guidance on this one. 

I still think infinite scroll is bad UX, if you know exactly what which visit you're looking for and it's near the end of the list. That said, even with pagination alone, the user needs to fish for the right page containing the visit they want. How about pagination + date picker?

ibacher commented:
So, my only reason for arguing for infinite scroll here is to keep things consistent with the Location Picker, since that's  what this most closely resembles. I definitely agree with a date picker or better yet, at date range picker.

denniskigen commented:
@ibacher @chibongho do you reckon this is good to go for this release cycle?

ibacher commented:
Yes, I think so.

chibongho commented:
Updated visit context switcher to use infinite scroll instead of pagination. Also added a date picker to limit visits started on or prior to selected date.

https://github.com/user-attachments/assets/736ed87f-c1f6-4069-8d98-3eb0892d2577

Based on (somewhat outdated) mockup from: 
https://app.zeplin.io/project/60d59321e8100b0324762e05/screen/679761bdffaf2ed81745f257

ibacher commented on packages/esm-patient-chart-app/src/visit/visits-widget/visit-context/visit-context-switcher.modal.tsx:
```diff
@@ -21,13 +23,27 @@ const VisitContextSwitcherModal: React.FC<VisitContextSwitcherProps> = ({
+  const rep = 'custom:(uuid,display,visitType,startDatetime,stopDatetime,location)';
+  const { visits, isLoading, error, hasMore, loadMore } = useInfiniteVisits(
+    patientUuid,
+    { toStartDate: dayjs(maxStartDateDebounced).endOf('day').toISOString() },
```
Shouldn't the filtering for startDate be up to the beginning of the day or an I misunderstanding the logic here?

chibongho replied:
The filter is for start times on or prior to that date, so if "2025-05-01" is selected, it should include all times prior to "2025-05-01 11:59:59 PM" in local time.

ibacher replied:
Got it!

ibacher commented on packages/esm-patient-chart-app/src/visit/visits-widget/visit.resource.tsx:
```diff
@@ -5,28 +5,45 @@ import {
+) {
   const url = new URL(
-    makeUrl(`${restBaseUrl}/visit?patient=${patientUuid}&v=${customRepresentation}`),
+    `${window.openmrsBase}/${restBaseUrl}/visit?patient=${patientUuid}&v=${customRepresentation}`,
```
It's preferential to use `getOpenmrsSpaBase()`, which just guarantees that there is always a `/` at the end of the `spaBase`. The `spaBase` variable itself is whatever was defined at build time, which may or may not have a trailing slash.

```suggestion
    `${window.getOpenmrsSpaBase()}${restBaseUrl}/visit?patient=${patientUuid}&v=${customRepresentation}`,
```

ibacher commented on packages/esm-patient-common-lib/src/pagination/pagination.component.tsx:
```diff
@@ -31,7 +34,13 @@ export const PatientChartPagination: React.FC<PatientChartPaginationProps> = ({
       {totalItems > 0 && (
-        <div className={isTablet ? styles.tablet : styles.desktop}>
+        <div
+          className={classNames(
+            isTablet ? styles.tablet : styles.desktop,
+            styles.paginationContainer,
+            grey ? styles.greyBackground : '',
+          )}
```
One reason to use `classnames` is because it supports this ternary thing kind of natively, e.g.

```suggestion
          className={classNames(styles.paginationContainer, {
          	[styles.tablet]: isTablet,
            [styles.desktop]: !isTablet ,
            [styles.greyBackground]: grey
          })}
```

ibacher reviewed (commented):
Couple of small suggestions and a question, but basically LGTM
